### PR TITLE
Universal header

### PR DESCRIPTION
**Pull request by viktor-yakubiv**

### CODE REVIEW

**Comment by viktor-yakubiv** [comment]
@Joozty I almost finished this. One ~small~ big thing left: develop and add the search field here.

File: src/modules/header/styles.css
Comment (Joozty) [inline src/modules/header/styles.css:210]: this can be probably removed, cannot be?

File: src/modules/header/styles.css
Comment (Joozty) [inline src/modules/header/styles.css:143]: this rule doesn't change anything, I think you can safely remove it

File: src/modules/header/styles.css
Comment (Joozty) [inline src/modules/header/styles.css:207]: It may overflow. Maybe you should break it sooner?

image

**Review by Joozty — APPROVED** [review]
Nice work. Please have a look at my comments. I didn't find anything blocking. All the things can be done (if needed) later on.  So you got my approval. 🙂 


@viktor-yakubiv What do you think about hiding menu automatically on focus out? it could be done with `:focus, :focus-within` (`:hover`) rules somehow. As I user I find it quite useful don't you think? :)

File: src/elements/card/card.md
Comment (viktor-yakubiv) [reply src/elements/card/card.md:1]: It's a leftover, I was playing with my broken syntax highlighting when some JSX appears in markdown code.

File: src/modules/header/styles.css
Comment (viktor-yakubiv) [reply src/modules/header/styles.css:113]: It should look like an arrow down. Doesn't it?

File: src/modules/header/header.jsx
Comment (viktor-yakubiv) [reply src/modules/header/header.jsx:100]: Why? Because it's not used? (perhaps, I don't remember)

File: src/modules/header/header.jsx
Comment (viktor-yakubiv) [reply src/modules/header/header.jsx:127]: If it gets integrated to oacore/next, it will break local development. I find such kind of URLs safe enough to go ahead. I didn't want to develop another complex solution for URL management here and kept it simple at the moment.

File: src/modules/header/styles.css
Comment (Joozty) [reply src/modules/header/styles.css:113]: it doesn't not for me. It looks like grey square because of this invalid rule.

File: src/modules/header/header.jsx
Comment (Joozty) [reply src/modules/header/header.jsx:100]: yes `main` doesn't exist

File: src/modules/header/header.jsx
Comment (viktor-yakubiv) [reply src/modules/header/header.jsx:8]: It's oacore/design, built with pure Babel and PostCSS. Importing will require adding some stuff to the build system for a single image. It's not worth it

File: src/modules/header/header.jsx
Comment (Joozty) [reply src/modules/header/header.jsx:127]: oh, that's true. What about basePath prop for header?

File: src/modules/header/styles.css
Comment (viktor-yakubiv) [reply src/modules/header/styles.css:207]: No, the menu is smaller than `768px`. The breakpoint is applied to the full width of the screen when the menu is visible only on part of it. It should be safe to leave it as is.

File: src/modules/header/header.jsx
Comment (viktor-yakubiv) [reply src/modules/header/header.jsx:127]: Is it really needed? It may get confusing soon if we add `/data-providers` or `/works` or something related to the oacore/search. I am afraid there is no better way to handle this than it's now.

File: src/modules/header/header.jsx
Comment (viktor-yakubiv) [reply src/modules/header/header.jsx:100]: You rely on some ID in Search for the portal search field. Should it be included here?

Comment (Joozty) [reply src/modules/header/header.jsx:100]: shoot I forgot about it. The problem is not with the id, that can be solved differently but with applying [classes](https://github.com/oacore/search/blob/master/templates/data-providers/styles.module.css#L10) to the item. Cannot the interface be slightly changed, i.e. any children passed down to header will be rendered in the middle?


```
<Header>
      <AppBar|Header.Item><Select>here</Select></AppBar|Header.Item>
</Header>
```

Comment (viktor-yakubiv) [reply src/modules/header/header.jsx:100]: I was afraid of that, because it may break the spacing there. However, it won't not bad, I think. Let's try it

File: src/modules/header/styles.css
Comment (viktor-yakubiv) [reply src/modules/header/styles.css:113]: I see. It works in Firefox Dev, Chrome Canary and Safari 14 but doesn't work for the stable Chrome. I will add a `@supports` rule for it.

File: src/modules/header/styles.css
Comment (viktor-yakubiv) [inline src/modules/header/styles.css:139]: This flex breaks with the SVG image in Safari.

File: src/modules/header/styles.css
Comment (viktor-yakubiv) [reply src/modules/header/styles.css:143]: It's safer to have it there. If it's not there, `auto` is set which is equal to regular content placement. It produces a similar result but may get broken if the markup changes.

File: src/modules/header/styles.css
Comment (viktor-yakubiv) [reply src/modules/header/styles.css:210]: Yes, it's a leftover, thank you 🙂

File: src/modules/header/styles.css
Comment (viktor-yakubiv) [reply src/modules/header/styles.css:14]: I get this but I don't know where it comes from

li::before problem

Comment (viktor-yakubiv) [reply src/modules/header/styles.css:14]: That's it: https://github.com/csstools/sanitize.css/blob/master/sanitize.css#L123-L129

I don't know now, should I keep it and find another trick or remove it...

File: src/modules/header/header.jsx
Comment (viktor-yakubiv) [reply src/modules/header/header.jsx]: I did it. Please, be aware, I am centring that element applying `margin-*: auto` to the logo and the nav.

File: src/modules/header/styles.css
Comment (Joozty) [inline src/modules/header/styles.css:76]: I think the menu needs some z-index and background colour

image

File: src/modules/header/styles.css
Comment (Joozty) [inline src/modules/header/styles.css:10]: This is something we should be probably getting rid off. I mean if I define a fixed height for the header the buttons are not stretched to the whole area.  And this height has to be defined because the main area should be moved down by that offset. (--app-bar-skip)

File: src/modules/header/header.jsx
Comment (Joozty) [inline src/modules/header/header.jsx:94]: please include `id`  here too :)

File: src/modules/header/styles.css
Comment (viktor-yakubiv) [reply src/modules/header/styles.css:10]: What do you mean saying 'get rid off'?

File: src/modules/header/styles.css
Comment (viktor-yakubiv) [reply src/modules/header/styles.css:76]: There is undocumented `--header-zindex` (or 1000 by default) applied for the desktop only. I will move this for mobile too.

Comment (viktor-yakubiv) [reply src/modules/header/styles.css:76]: I see the problem: if `position: fixed` is applied to the overall component, you lack the background and z-index. I am afraid, `z-index` should be applied manually.

Comment (viktor-yakubiv) [reply src/modules/header/styles.css:76]: Also, you can wonderfully drop `fixed` prop there which gets automatically passed to the `AppBar` and solves your problem.

**Comment by viktor-yakubiv** [comment]
@Joozty I guess you had problems because I didn't push everything. Would you like to check it again? If not, I can rebase and merge this one, and we keep going

File: src/modules/header/styles.css
Comment (Joozty) [reply src/modules/header/styles.css:76]: I cannot drop position fixed because when you are on the mobile menu and open the submenu the main content gets moved down and I don't think it should work like that

File: src/modules/header/styles.css
Comment (Joozty) [reply src/modules/header/styles.css:10]: I mean if we have it there than buttons are not going to be stretched properly when you specify a height for app bar and imho it should by default

image

File: src/modules/header/styles.css
Comment (viktor-yakubiv) [reply src/modules/header/styles.css:76]: > There is undocumented `--header-zindex` (or 1000 by default) applied for the desktop only. I will move this for mobile too.

Renamed to `--header-z-index` and defaults to `--app-bar-z-index + 10`.

---

I don't know, perhaps not. But if you make the app-bar fixed, it won't do so, child menus will overlay the content of the page. I think it's good option.

File: src/modules/header/styles.css
Comment (viktor-yakubiv) [reply src/modules/header/styles.css:10]: Height for the AppBar? Why do you need it? I was trying to respect spacing as much as possible. The AppBar should be rendered in a good shape without height specified.

Comment (Joozty) [reply src/modules/header/styles.css:10]: That's how our app structure works in search as well as in dashboard (not relevant). And the reasons I can think of only one at the moment and I already described it here https://github.com/oacore/design/pull/358#discussion_r524409693 but there were probably more reasons?

Comment (Joozty) [reply src/modules/header/styles.css:10]: @viktor-yakubiv Any update on this? What's your decision here? I see this as the last blocker for merging this PR.

Comment (viktor-yakubiv) [reply src/modules/header/styles.css:10]: I finally got what's your problem is. Sorry, I did not get it before. This is how it looks for me:

Screenshot 2020-11-23 at 11 15 54

I will have a look at what I can do with `position: relative`

Comment (viktor-yakubiv) [reply src/modules/header/styles.css:10]: Please, have a look at the [solution](https://github.com/oacore/design/pull/358/commits/326c2f95a255b7cb6d784aba5f1cfc7681a6cb3d). Doesn't it feel bug-causing?

Comment (Joozty) [reply src/modules/header/styles.css:10]: I hope it's ok. Please remove `FIXME` comment, because it will stay there forever. 😄 and after that feel free to merge. Thanks a lot for your work :)

**Pull request merged**
